cpp/src/arrow/compute/kernels/scalar_cast_nested.cc

```diff
@@ -404,14 +400,20 @@ struct CastStruct {
 
     int out_field_index = 0;
     for (int field_index : fields_to_select) {
-      const auto& target_type = out->type()->field(out_field_index++)->type();
+      const auto& target_field = out->type()->field(out_field_index++);
+      const auto& target_type = target_field->type();
       if (field_index == kFillNullSentinel) {
         ARROW_ASSIGN_OR_RAISE(auto nulls,
                               MakeArrayOfNull(target_type->GetSharedPtr(), batch.length));
         out_array->child_data.push_back(nulls->data());
       } else {
         const auto& values = (in_array.child_data[field_index].ToArrayData()->Slice(
             in_array.offset, in_array.length));
+        if (!target_field->nullable() && values->null_count > 0) {
+          return Status::TypeError("field '", target_field->name(),
```

Comment:
Besides, should we discard the `buffer[0]` if new data doesn't contains any nulls?